packages/salesforcedx-vscode-core/src/commands/forceSourceDeploy.ts

```diff
@@ -97,9 +100,35 @@ export class ForceSourceDeployExecutor extends SfdxCommandletExecutor<
   }
 }
 
+export class MultipleSourcePathsGatherer
+  implements ParametersGatherer<SelectedPath> {
+  private uris: vscode.Uri[];
+  public constructor(uris: vscode.Uri[]) {
+    this.uris = uris;
+  }
+  public async gather(): Promise<ContinueResponse<SelectedPath>> {
+    const sourcePaths = this.uris.map(uri => uri.fsPath).join(',');
+    return {
+      type: 'CONTINUE',
+      data: { filePath: sourcePaths, type: FileType.Source }
+    };
+  }
+}
+
 const workspaceChecker = new SfdxWorkspaceChecker();
 
-export async function forceSourceDeploy(explorerPath: any) {
+export async function forceSourceDeployMultipleSourcePaths(uris: vscode.Uri[]) {
```

Comment:
This new deploy command is only called when a file within an sfdx packageDirectory is created/changed/deleted so it does not need to use the ManifestOrSourcePathGatherer to determine whether the passed in paths belong to source or manifests.